go/store/nbs/archive_reader.go

```diff
@@ -0,0 +1,382 @@
+// limitations under the License.
+
+package nbs
+
+import (
+	"bufio"
+	"crypto/sha512"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"math/bits"
+
+	"github.com/dolthub/gozstd"
+	"github.com/pkg/errors"
+
+	"github.com/dolthub/dolt/go/store/hash"
+)
+
+type archiveReader struct {
+	reader    io.ReaderAt
+	prefixes  []uint64
+	byteSpans []byteSpan
+	chunkRefs []chunkRef
+	suffixes  []suffix
+	footer    footer
+}
+
+type chunkRef struct {
+	dict uint32
+	data uint32
+}
+
+type suffix [hash.SuffixLen]byte
+
+type footer struct {
+	indexSize     uint32
+	byteSpanCount uint32
+	chunkCount    uint32
+	metadataSize  uint32
+	dataCheckSum  sha512Sum
+	indexCheckSum sha512Sum
+	metaCheckSum  sha512Sum
+	formatVersion byte
+	fileSignature string
+	fileSize      uint64 // Not actually part of the footer, but necessary for calculating offsets.
+}
+
+// dataSpan returns the span of the data section of the archive. This is not generally used directly since we usually
+// read individual spans for each chunk.
+func (f footer) dataSpan() byteSpan {
+	return byteSpan{offset: 0, length: f.fileSize - archiveFooterSize - uint64(f.metadataSize) - uint64(f.indexSize)}
+}
+
+// totalIndexSpan returns the span of the entire index section of the archive. This span is not directly useful as
+// the index is broken into a compressed section and an uncompressed section. Use indexCompressedSpan and indexSuffixSpan
+func (f footer) totalIndexSpan() byteSpan {
+	return byteSpan{offset: f.fileSize - archiveFooterSize - uint64(f.metadataSize) - uint64(f.indexSize), length: uint64(f.indexSize)}
+}
+
+// indexCompressedSpan returns the span of the index section of the archive.
+func (f footer) indexCompressedSpan() byteSpan {
+	suffixLen := uint64(f.chunkCount * hash.SuffixLen)
+	totalIdx := f.totalIndexSpan()
+	return byteSpan{offset: totalIdx.offset, length: totalIdx.length - suffixLen}
+}
+
+func (f footer) indexSuffixSpan() byteSpan {
+	suffixLen := uint64(f.chunkCount * hash.SuffixLen)
+	totalIdx := f.totalIndexSpan()
+	compressedLen := totalIdx.length - suffixLen
+
+	return byteSpan{totalIdx.offset + compressedLen, suffixLen}
+}
+
+func (f footer) metadataSpan() byteSpan {
+	return byteSpan{offset: f.fileSize - archiveFooterSize - uint64(f.metadataSize), length: uint64(f.metadataSize)}
+}
+
+func newArchiveReader(reader io.ReaderAt, fileSize uint64) (archiveReader, error) {
+	footer, err := loadFooter(reader, fileSize)
+	if err != nil {
+		return archiveReader{}, err
+	}
+
+	indexSpan := footer.indexCompressedSpan()
+	secRdr := io.NewSectionReader(reader, int64(indexSpan.offset), int64(indexSpan.length))
+	rawReader := bufio.NewReader(secRdr)
+
+	errChan := make(chan error, 1)
+
+	redr, wrtr := io.Pipe()
+	go func() {
+		defer wrtr.Close()
+		err := gozstd.StreamDecompress(wrtr, rawReader)
+		if err != nil {
+			errChan <- errors.Wrap(err, "Failed to decompress archive index")
+		}
+
+		close(errChan)
+	}()
```

Comment:
I think is simpler and maybe more idiomatic with something like:

```go
go func() {
    err := gozstd.StreamDecompress(wrtr, rawReader)
    if err != nil {
        wrtr.CloseWithError(err)
    } else {
        wrtr.Close()
    }
}()
```

Then the reader will just see an error if `StreamDecompress` errors...otherwise it will see EOF as success.